14_ComputeReduce/main.cpp

```diff
@@ -0,0 +1,281 @@
+		reducer->updateDescriptorSet(m_device.get(), ds.get(), in_gpu_range, scratch_gpu_range);
+		
+		{
+			smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(computeQueue->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+			if (!cmdpool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, { &cmdbuf,1 }))
+			{
+				logFail("Failed to create Command Buffers!\n");
+				return false;
+			}
+		}
+
+		cmdbuf->begin(IGPUCommandBuffer::USAGE::SIMULTANEOUS_USE_BIT); // (REVIEW): not sure about this
+		cmdbuf->fillBuffer(scratch_gpu_range, 0u);
+		cmdbuf->bindComputePipeline(reduce_pipeline);
+		auto pipeline_layout = reduce_pipeline->getLayout();
+		cmdbuf->bindDescriptorSets(asset::EPBP_COMPUTE, pipeline_layout, 0u, 1u, &ds.get());
+		reducer->dispatchHelper(cmdbuf.get(), pipeline_layout, reduce_push_constants, reduce_dispatch_info, 0u, nullptr, 0u, nullptr);
+		cmdbuf->end();
+
+		{
+			semInfo[0].value = 2;
+			semInfo[0].stageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			const IQueue::SSubmitInfo::SCommandBufferInfo commandBuffers[1] = { {
+				.cmdbuf = cmdbuf.get()
+			} };
+				
+			const IQueue::SSubmitInfo infos[1] = { {
+				.commandBuffers = commandBuffers,
+				.signalSemaphores = semInfo
+			} };
+
+			computeQueue->startCapture();
+			if (computeQueue->submit(infos) != IQueue::RESULT::SUCCESS) {
+				m_logger->log("Submission failure", system::ILogger::ELL_ERROR);
+			}
+			computeQueue->endCapture();
+		}
+
+		// TODO: Update to support all operations
+		// cpu counterpart
+		auto cpu_begin = inputData + begin;
+		m_logger->log("CPU reduce begin", system::ILogger::ELL_PERFORMANCE);
+
+		auto start = std::chrono::high_resolution_clock::now();
+		auto result = std::reduce(cpu_begin, inputData + end, 0u);
+		auto stop = std::chrono::high_resolution_clock::now();
+
+		m_logger->log("CPU reduce end. Time taken: %d us", system::ILogger::ELL_PERFORMANCE, std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count());
+
+		// wait for the gpu impl to complete
+		const ISemaphore::SWaitInfo cmdbufDonePending[] = {{
+			.semaphore = semaphore.get(),
+			.value = 2
+		}};
+		if (m_device->blockForSemaphores(cmdbufDonePending) != ISemaphore::WAIT_RESULT::SUCCESS) {
+			m_logger->log("Blocking for operation semaphore failed", ILogger::ELL_ERROR);
+			return false;
+		}
+		
+		{
+			IGPUBuffer::SCreationParams params = {};
+			params.size = in_gpu_range.size;
+			params.usage = IGPUBuffer::EUF_TRANSFER_DST_BIT;
+			// (REVIEW): Check if this new download_buffer is needed or if we can directly read from the gpu_input buffer
+			auto downloaded_buffer = m_device->createBuffer(std::move(params));
+			auto memReqs = downloaded_buffer->getMemoryReqs();
+			memReqs.memoryTypeBits &= m_physicalDevice->getDownStreamingMemoryTypeBits();
+			auto queriesMem = m_device->allocate(memReqs, downloaded_buffer.get());
```

Comment:
use the downstreaming buffer and make a consumption callback.

Otherwise you'll have a problem making a 128MB HOST_VISIBLE and DEVICE_LOCAL buffer, not everyone has ReBAR configured correctly.

Reply:
See example 05 (whichever does the streaming consumer callback), how to download a buffer and consume it "piecewise"